### PR TITLE
Truncate k8s hostnames to be no longer than 63 characters

### LINKED ISSUE #101: Pod hostnames truncated to 63 characters
I'm seeing this exception in an executor pod:

```
2017-02-10 01:05:38 WARN  BlockManager:87 - Failed to fetch remote block broadcast_0_piece0 from BlockManagerId(2, org-apache-spark-examples-sparkpi-1486688665059-exec-a365b3dc4b, 32787, None) (failed attempt 1)
org.apache.spark.SparkException: Exception thrown in awaitResult:
        at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:205)
        at org.apache.spark.network.BlockTransferService.fetchBlockSync(BlockTransferService.scala:104)
        at org.apache.spark.storage.BlockManager.getRemoteBytes(BlockManager.scala:579)
        at org.apache.spark.broadcast.TorrentBroadcast$$anonfun$org$apache$spark$broadcast$TorrentBroadcast$$readBlocks$1.apply$mcVI$sp(TorrentBroadcast.scala:161)
        at org.apache.spark.broadcast.TorrentBroadcast$$anonfun$org$apache$spark$broadcast$TorrentBroadcast$$readBlocks$1.apply(TorrentBroadcast.scala:150)
        at org.apache.spark.broadcast.TorrentBroadcast$$anonfun$org$apache$spark$broadcast$TorrentBroadcast$$readBlocks$1.apply(TorrentBroadcast.scala:150)
        at scala.collection.immutable.List.foreach(List.scala:381)
        at org.apache.spark.broadcast.TorrentBroadcast.org$apache$spark$broadcast$TorrentBroadcast$$readBlocks(TorrentBroadcast.scala:150)
        at org.apache.spark.broadcast.TorrentBroadcast$$anonfun$readBroadcastBlock$1.apply(TorrentBroadcast.scala:222)
        at org.apache.spark.util.Utils$.tryOrIOException(Utils.scala:1273)
        at org.apache.spark.broadcast.TorrentBroadcast.readBroadcastBlock(TorrentBroadcast.scala:206)
        at org.apache.spark.broadcast.TorrentBroadcast._value$lzycompute(TorrentBroadcast.scala:66)
        at org.apache.spark.broadcast.TorrentBroadcast._value(TorrentBroadcast.scala:66)
        at org.apache.spark.broadcast.TorrentBroadcast.getValue(TorrentBroadcast.scala:96)
        at org.apache.spark.broadcast.Broadcast.value(Broadcast.scala:70)
        at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:82)
        at org.apache.spark.scheduler.Task.run(Task.scala:114)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:313)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: Failed to connect to org-apache-spark-examples-sparkpi-1486688665059-exec-a365b3dc4b:32787
        at org.apache.spark.network.client.TransportClientFactory.createClient(TransportClientFactory.java:230)
        at org.apache.spark.network.client.TransportClientFactory.createClient(TransportClientFactory.java:181)
        at org.apache.spark.network.netty.NettyBlockTransferService$$anon$1.createAndStart(NettyBlockTransferService.scala:97)
        at org.apache.spark.network.shuffle.RetryingBlockFetcher.fetchAllOutstanding(RetryingBlockFetcher.java:140)
        at org.apache.spark.network.shuffle.RetryingBlockFetcher.access$200(RetryingBlockFetcher.java:43)
        at org.apache.spark.network.shuffle.RetryingBlockFetcher$1.run(RetryingBlockFetcher.java:170)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144)
        ... 1 more
Caused by: java.nio.channels.UnresolvedAddressException
        at sun.nio.ch.Net.checkAddress(Net.java:101)
        at sun.nio.ch.SocketChannelImpl.connect(SocketChannelImpl.java:622)
        at io.netty.channel.socket.nio.NioSocketChannel.doConnect(NioSocketChannel.java:242)
        at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.connect(AbstractNioChannel.java:205)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.connect(DefaultChannelPipeline.java:1226)
        at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:550)
        at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:535)
        at io.netty.channel.ChannelOutboundHandlerAdapter.connect(ChannelOutboundHandlerAdapter.java:47)
        at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:550)
        at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:535)
        at io.netty.channel.ChannelDuplexHandler.connect(ChannelDuplexHandler.java:50)
        at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:550)
        at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:535)
        at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:517)
        at io.netty.channel.DefaultChannelPipeline.connect(DefaultChannelPipeline.java:970)
        at io.netty.channel.AbstractChannel.connect(AbstractChannel.java:215)
        at io.netty.bootstrap.Bootstrap$2.run(Bootstrap.java:166)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:408)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:455)
        at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:140)
        ... 2 more
```

where it can't look up the hostname `org-apache-spark-examples-sparkpi-1486688665059-exec-a365b3dc4b`

The pod names look like this though:
```
$ kubectl get pods
NAME                                                                                    READY     STATUS    RESTARTS   AGE
org-apache-spark-examples-sparkpi-1486688665059                                         1/1       Running   0          2m
org-apache-spark-examples-sparkpi-1486688665059-exec-4d0790652fdd47afb530de07df5fddb9   1/1       Running   0          2m
org-apache-spark-examples-sparkpi-1486688665059-exec-a365b3dc4b5d4aa6a4250fd472720493   1/1       Running   0          2m
org-apache-spark-examples-sparkpi-1486688665059-exec-d5141f09a53c4642a621a6974152835d   1/1       Running   0          2m
org-apache-spark-examples-sparkpi-1486688665059-exec-eecf79463a934db4adb3d86dbfd46589   1/1       Running   0          2m
org-apache-spark-examples-sparkpi-1486688665059-exec-f2c7ea5037104aa5aac2b0ff1c2dbff4   1/1       Running   0          2m
$
```

note that the suffix is much longer in the pod names than in the hostname being queried from inside the executor.

Looking at the relative lengths:

```
$ echo -n org-apache-spark-examples-sparkpi-1486688665059-exec-eecf79463a934db4adb3d86dbfd46589 | wc
      0       1      85
$ echo -n org-apache-spark-examples-sparkpi-1486688665059-exec-a365b3dc4b | wc
      0       1      63
$
```

it looks like something is truncating the peer executor's hostname to 63 characters.

Googling around, it looks like the maximum length of a hostname (just one part of a FQDN) is 63 characters: http://stackoverflow.com/questions/8724954/what-is-the-maximum-number-of-characters-for-a-host-name-in-unix

I think we need to ensure by whatever means that the hostnames of the pods we create are under 63 characters so we don't get these various issues with less-supported long hostnames.

Similar to https://github.com/apache-spark-on-k8s/spark/issues/82 but this issue is for hostname truncation, whereas that one had DNS failures despite no hostname truncation.

### PR DESCRIPTION
**Pull request by ash211**
Fixes #101

### CODE REVIEW

**Comment by foxish** [comment]
I see, the executors use FQDNs and not IP addresses to communicate with each other? I was expecting that we'd only need to truncate the hostnames of pods we want to expose via a service.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala [hunk 1]
```diff
@@ -157,7 +157,13 @@ private[spark] class KubernetesClusterSchedulerBackend(
   private def allocateNewExecutorPod(): (String, Pod) = {
     val executorKubernetesId = UUID.randomUUID().toString.replaceAll("-", "")
```
Comment (foxish) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala]: What is this UUID being used for now?

Comment (ash211) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala]: It's used as the identifier for each executor that the `KubernetesClusterSchedulerBackend` keeps in its `runningExecutorPods` map (identifier -> Pod).

Then when removing executors in `doKillExecutors` this is the identifier `CoarseGrainedSchedulerBackend` uses to identify which executors to be killed.

However, I don't see where the executorId we receive in `doKillExecutors` is ever given to the rest of Spark or the `CoarseGrainedSchedulerBackend`.  @mccheah is this a remnant of earlier versions where we aimed to support dynamic allocation?

**Review by ash211 — COMMENTED** [review]
You're right -- currently the communication from pod to pod happens via FQDNs, not IP addresses.  In YARN and Mesos this is the standard, but it's very possible that we want IP addresses in k8s given that it's so IP-centric.

Matt what do you think about running executor -> executor comms via IPs rather than via FQDNs?

**Review by foxish — APPROVED** [review]
LGTM

**Comment by ash211** [comment]
@mccheah any objections?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala (hunk 1 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala]: See `CoarseGrainedSchedulerBackend#receiveAndReply()` - the `RegisterExecutor` case.

**Comment by iyanuobidele** [comment]
This looks like a straightforward change to me. 
Although, if the appId (supplied by the user) is long enough to make the 63 char end at `...-exec`, then we might have a problem. But, I don't know how practical it is to have such long appIds.

**Comment by mccheah** [comment]
Fine with truncating. Follow up discussion may be needed to consider IP over FQDN, but I lean towards using IP addresses .

**Pull request merged**